### ISSUE TITLE
Question about pre-training BEiT-base on ImageNet-1k and then fine-tuning on ADE20K

### ISSUE THREAD

**Issue by laojiangwei** [microsoft/unilm#391]
**Describe**
Model I am using (UniLM, MiniLM, LayoutLM ...): BEIT
I am try to reproducing self-supervised pre-training BEiT-base on ImageNet-1k and then fine-tuning on ADE20K, in your paper it will get mIoU 45.6, slightly higher than Supervised Pre-Training on ImageNet(45.3) and DINO(44.1), But i can not reproduce this result, i only get mIoU about 39.

I wonder whether you can provide all the training commands and hyperparams for self-supervised pre-training BEiT-base on ImageNet-1k and then fine-tuning on ADE20K. Thanks~

**Comment by donglixp** [comment]
@laojiangwei  Could you provide more detailed commands of your runs?

**Comment by laojiangwei** [comment]
the training commands for self-supervised pre-training BEiT-base on ImageNet-1k :
OMP_NUM_THREADS=1 python -m torch.distributed.launch --nproc_per_node=8 run_beit_pretraining.py \
        --data_path ${DATA_PATH} --output_dir ${OUTPUT_DIR} --num_mask_patches 75 \
        --model beit_base_patch16_224_8k_vocab --discrete_vae_weight_path ${TOKENIZER_PATH} \
        --batch_size 64 --lr 1.5e-3 --warmup_epochs 10 --epochs 300 \
        --clip_grad 3.0 --drop_path 0.1 --layer_scale_init_value 0.1 \
        --imagenet_default_mean_and_std

the training commands for fine-tuning on ADE20K:
bash tools/dist_train.sh configs/beit/upernet/upernet_beit_base_12_512_slide_160k_ade20k.py 8 --work-dir work_dirs --seed 0  --deterministic --options model.pretrained=${checkpoint of self-supervised pre-training BEiT-base on ImageNet-1k}

the training commands are default, i only reduced batchsize of self-supervised pre-training BEiT-base on ImageNet-1k.

**Comment by laojiangwei** [comment]
> @laojiangwei Could you provide more detailed commands of your runs?

the training commands for self-supervised pre-training BEiT-base on ImageNet-1k :
OMP_NUM_THREADS=1 python -m torch.distributed.launch --nproc_per_node=8 run_beit_pretraining.py
--data_path ${DATA_PATH} --output_dir ${OUTPUT_DIR} --num_mask_patches 75
--model beit_base_patch16_224_8k_vocab --discrete_vae_weight_path ${TOKENIZER_PATH}
--batch_size 64 --lr 1.5e-3 --warmup_epochs 10 --epochs 300
--clip_grad 3.0 --drop_path 0.1 --layer_scale_init_value 0.1
--imagenet_default_mean_and_std

the training commands for fine-tuning on ADE20K:
bash tools/dist_train.sh configs/beit/upernet/upernet_beit_base_12_512_slide_160k_ade20k.py 8 --work-dir work_dirs --seed 0 --deterministic --options model.pretrained=${checkpoint of self-supervised pre-training BEiT-base on ImageNet-1k}

the training commands are default, i only reduced batchsize of self-supervised pre-training BEiT-base on ImageNet-1k.

**Comment by donglixp** [comment]
@addf400 would help to double check the above commands.

**Comment by addf400** [comment]
@laojiangwei , Can you share the mlm_acc and mlm_loss of the final epoch?
I noticed that you keep the same learning rate 1.5e-3 while decrease the batch size from 2048 to 512.

**Comment by donglixp** [comment]
@laojiangwei The LR can be reduced by 2x if you would like to use bsz=512.

**Comment by laojiangwei** [comment]
@addf400 @donglixp I try to train self-supervised pre-training BEiT-base on ImageNet-1k again, this time i use batchsize 2048 as default. the training commands for self-supervised pre-training BEiT-base on ImageNet-1k :

OMP_NUM_THREADS=1 python -m torch.distributed.launch --nproc_per_node=8 --nnodes 2 --node_rank $RANK --master_port $MASTER_PORT --master_addr $MASTER_ADDR run_beit_pretraining.py \
        --data_path ${DATA_PATH} --output_dir ${OUTPUT_DIR} --num_mask_patches 75 \
        --model beit_base_patch16_224_8k_vocab --discrete_vae_weight_path ${TOKENIZER_PATH} \
        --batch_size 128 --lr 1.5e-3 --warmup_epochs 10 --epochs 300 \
        --clip_grad 3.0 --drop_path 0.1 --layer_scale_init_value 0.1 \
        --imagenet_default_mean_and_std

the mlm_acc and mlm_loss of the final epoch is :

{"train_lr": 1.001460655613469e-05, "train_min_lr": 1.001460655613469e-05, "train_mlm_acc": 0.04951678842902184, "train_loss": 6.301707845926285, "train_loss_scale": 6363.5456, "train_weight_decay": 0.050000000000000495, "train_grad_norm": NaN, "epoch": 299, "n_parameters": 91965776}

the result of fine-tuning on ADE20K is:

{"mode": "val", "epoch": 127, "iter": 160000, "lr": 0.0, "mIoU": 0.4379, "mAcc": 0.5517, "aAcc": 0.8178}

Is this a normal result? I found that there was still a gap with mIoU 45.6

**Comment by addf400** [comment]
@laojiangwei The hyper-parameters in this [config]( https://github.com/microsoft/unilm/blob/master/beit/semantic_segmentation/configs/beit/upernet/upernet_beit_base_12_512_slide_160k_ade20k.py) is used for the BEiT models that are intermediate-fine-tuned on ImageNet-22k. When directly using the pretrained checkpoints, there were some differences in terms of hyperparams, such as larger learning rate. We will also add the commands to the repo for your current setup soon. Thank you for the feedback.

**Comment by addf400** [comment]
@laojiangwei Please try this [config](https://github.com/microsoft/unilm/blob/master/beit/semantic_segmentation/configs/beit/upernet/upernet_beit_base_12_512_slide_160k_ade20k_pt.py). 
In our ablation experiment, the performance can also be further improved by increasing the number of pretraining epochs from 300 to 800.

**Comment by jizxny** [comment]
@addf400 Hi, I'm wondering 300e pretraining result on ADE20k 44.65 is achieved with SETR-PUP head or UperNet head. It seems these two are not identical in my unserstands.

**Comment by donglixp** [comment]
> @addf400 Hi, I'm wondering 300e pretraining result on ADE20k 44.65 is achieved with SETR-PUP head or UperNet head. It seems these two are not the same in my understanding.

@jizxny The v1 arxiv preprint (https://arxiv.org/abs/2106.08254v1) used SETR-PUP. In the current github release, we switched to UperNet because its performance seems better.

**Comment by donglixp** [comment]
The code and pre-trained models of BEiT-3 can be found at [aka.ms/beit3](https://aka.ms/beit3).
